### ISSUE TITLE
`[tool.uv.sources]` can't be used as `[sources]` in `uv.toml`

### ISSUE THREAD

**Issue by johnpyp** [astral-sh/uv#6772]
As mentioned in the docs, `tool.uv.sources` works fine for specifying sources overrides, but it seems that `[sources]` in `uv.toml` does not.

Confusingly, using `tool.uv.sources` and a `uv.toml` *doing other things* throws this warning:
```
warning: Found both a `uv.toml` file and a `[tool.uv]` section in an adjacent `pyproject.toml`. The `[tool.uv]` section will be ignored in favor of the `uv.toml` file.
```

However, it actually works fine to have it split among the two files.

Let me know if you need more to reproduce, but it should be straightforward as it's just a config file distinction for all forms of `[tool.uv.sources]`.

---
**Version:** uv 0.3.3
**Platform**: MacOS / Linux
**Command**: `uv sync`

**Comment by charliermarsh** [comment]
Can you share an example of the exact setup you're trying to achieve?

**Comment by charliermarsh** [comment]
I believe sources need to be in the `pyproject.toml` alongside the dependencies they annotate.

**Comment by johnpyp** [comment]
> I believe sources need to be in the pyproject.toml alongside the dependencies they annotate.

Yeah totally, this makes sense generally. The two things that are weird are:

1. it seems like `sources` is uniquely inconsistent with the rest of the `tool.uv` namespace, in that it doesn't support configuration in `uv.toml`, whereas basically every other property does.
2. Even if I keep `tool.uv.sources` in `pyproject.toml` but put the rest into `uv.toml`, it still gives that warning.

I'm just using `sources` for a normal path dependency, nothing special, but I was using `uv.toml` as my configuration for `index-url` and `extra-index-url`. These are urls generated on-demand, not static, so I don't want to put them into source control. Of course, that's another issue because I can't mix `pyproject.toml` and `uv.toml` generally, but that was the impotence to start using `uv.toml` in the first place.

**Comment by charliermarsh** [comment]
@colinjc -- Can you not put that configuration in a `pyproject.toml`?

**Comment by johnpyp** [comment]
Echoing Colin's usecase - generated, often not source-controlled config is very common in large environments for things like registry configuration that may not be static.

**Comment by charliermarsh** [comment]
I believe we intentionally don't support this and it will now error (instead of silently ignoring).

**Comment by befelix** [comment]
> I believe we intentionally don't support this and it will now error (instead of silently ignoring).

I would have another use-case for `sources` in `uv.toml`: My laptop doesn't have a GPU, so I generally want any `pytorch` package to be the cpu-only variant. While we can handle this for our local projects via `pyproject.toml`, if I install external dependencies directly into a fresh venv I have to manually pass the corresponding index url again. Adding the pytorch cpu-only index (without marking it at explicit) is not an option, since it has unintended side-effects unless we use an unsafe index strategy. Being able to explicitly pin `pytorch` to the cpu index globally via a `sources` field in the `uv.toml` would resolve that issue.

**Comment by ipeevski** [comment]
Along the lines of what has been said before, this is my use case (that gives the warning anytime I run `uv`):

I want to specify torch sources/indexes (I don't mind where, preferrably in the `uv.toml` file, but currently it only allows in `project.toml`)
I want to specify my `cache-dir` to be on the same drive as my project (so hard links work)
I don't want to commit potentially sensitive information about where my cache-dir is and more to the point, it will not be a valid location for most people (as not everyone has a P: drive). 

So, currently, I don't see a way to reconcile those two requirements (my wants above)

**Comment by charliermarsh** [comment]
Why can’t you put the cache-dir setting in a user-level uv.toml?

**Comment by orishamir** [comment]
> I believe we intentionally don't support this and it will now error (instead of silently ignoring).

May I add, that while having `[sources]` inside `uv.toml` of the project indeed yields an error stating `sources` field is not supported in a `uv.toml` file, it does not error in the case where `[sources]` appear in a `uv.toml` of system-level configuration (in _%SYSTEMDRIVE%\ProgramData\uv\uv.toml_ in windows at least).   
So the `[sources]` inside system-level configuration silently gets ignored.
A warning in that case would be helpful since it caused significant headaches 

_By the way - you are doing an incredible job with UV, I can't thank you enough!_

**Comment by cylixlee** [comment]
My usecase is very close to what @befelix describes.

I'm writing a lib based on PyTorch on my laptop, which does not have GPU, and I configured the `[tool.uv.sources]` section with an explicit `pytorch-cpu` index. However, because the `[tool.uv.sources]` section can only be specified in `pyproject.toml`, I have to **comment out the section every time** I publish it to PyPI.

If the `[sources]` section is allowed in `uv.toml`, I can easily use `pytorch-cpu` index locally -- I can add `uv.toml` to `.gitignore` and there's no need to comment out those configuration again!

uv is the best package manager I've ever met and I appreciate your work so much, and I can't be too grateful if this feature is accepted😊.

**Comment by konstin** [comment]
@cylixlee Have you seen https://docs.astral.sh/uv/guides/integration/pytorch/#configuring-accelerators-with-environment-markers? We added this feature specifically for that torch use case.

**Comment by cylixlee** [comment]
@konstin Yeah, the markers is a good choice, however it may influence downstream: I personally use Windows+PyTorch CPU and Linux+GPU, however the downstream may not want to pair up like that...

I also tried the extras, but writing a lot of optional dependencies for every PyTorch-related lib is not a very pleasant work.😂

**Comment by charliermarsh** [comment]
(If you use `uv run` in lieu of `uv sync`, it will never remove any installed extras.)

**Comment by rkechols** [comment]
I'd also like to put a vote in for being able to specify `tool.uv.sources` config outside of `pyproject.toml`. My use-case is also pytorch; some developer machines have GPUs, some don't, and it's independent of OS. To me the intuitive solution was to put `[[index]]` and `[sources]` entries in my `uv.toml` file, and I was surprised that the `[sources]` config was rejected.

Do we have any news whether this is on the roadmap?

**Comment by zanieb** [comment]
It's not on the roadmap, no. Defining sources outside the `pyproject.toml` would invalidate the lockfile. We'd rather invest in things like being able to toggle sources based on the presence of a GPU.

**Comment by ncoghlan** [comment]
I believe I just ran into this restriction trying to enable `uv` source overrides in `venvstacks`. That uses `uv pip compile` and `uv pip install` from input requirements files, with nary a `pyproject.toml` in sight.

There isn't any lock file to inadvertently invalidate in that usage model.

**Comment by zanieb** [comment]
That's fair, but allowing it in that context but not other ones adds a bunch of complexity! If we choose to do so, we'll need to be careful about it.

**Comment by ncoghlan** [comment]
Would it be viable to add a `pip.sources` table to constrain the non-project source definitions to just the use cases that aren't using `uv.lock`? (although that may be the extra complexity you're referring to) 

Independently of that question, I'll see if I can rework `venvstacks` to use pseudo-projects instead of requirements files.

**Comment by ncoghlan** [comment]
@charliermarsh In relation to https://github.com/astral-sh/uv/issues/6772#issuecomment-2487042845, does "it will now error" refer to attempting to specify `[sources]` in `uv.toml` at all, or more specifically to attempting to install a package that has an override specified in `uv.toml`?

Or is it only when `pyproject.toml` exists (and has a `tools.uv.sources` field?), so `uv pip compile` will never trigger it?

**Comment by TheTerrasque** [comment]
> It's not on the roadmap, no. Defining sources outside the `pyproject.toml` would invalidate the lockfile. We'd rather invest in things like being able to toggle sources based on the presence of a GPU.

I just came over this problem and I believe this is unwise. The sources settings in relation to pytorch is clearly a system-wide and not project setting. You not only have gpu / cpu, but type of gpu and with nvidia version of cuda installed. Furthermore, as some have pointed out here there's many situations where it's organization wide settings and again not project-scope settings.

**Comment by patrontheo** [comment]
> My usecase is very close to what [@befelix](https://github.com/befelix) describes.
> 
> I'm writing a lib based on PyTorch on my laptop, which does not have GPU, and I configured the `[tool.uv.sources]` section with an explicit `pytorch-cpu` index. However, because the `[tool.uv.sources]` section can only be specified in `pyproject.toml`, I have to **comment out the section every time** I publish it to PyPI.
> 
> If the `[sources]` section is allowed in `uv.toml`, I can easily use `pytorch-cpu` index locally -- I can add `uv.toml` to `.gitignore` and there's no need to comment out those configuration again!
> 
> uv is the best package manager I've ever met and I appreciate your work so much, and I can't be too grateful if this feature is accepted😊.

I have exactly the same use-case as @cylixlee and was hoping for the same solution (put the sources in `uv.toml` so that it applies locally only). If not possible in `uv.toml`, could it be in something like `pyproject.dev.toml`, that would get merged with the main pyproject when running `uv sync` ? Or do you see any other solution for that issue ? @zanieb

**Comment by Ruhrpottpatriot** [comment]
Adding to the people who would like this feature added.
I work in a company where we have some strict requirements which machine can access our internal network. However the projects I work on are sometimes developed in the field (quite literally) on machines which don't have access to the internal network. At that point we need to use local copies of our in-house packages.

Having the option to use the [sources] section in a uv.toml configuration would allow us to configure the machines we take with us to always use local copies and machines that are internal to always use our internal package registry. Much less hassle and much less error prone.

**Comment by zanieb** [comment]
>  The sources settings in relation to pytorch is clearly a system-wide and not project setting. 

We already allow markers on sources, which are for system-wide changes https://docs.astral.sh/uv/concepts/projects/dependencies/#platform-specific-sources

> Having the option to use the [sources] section in a uv.toml configuration would allow us to configure the machines we take with us to always use local copies and machines that are internal to always use our internal package registry. 

Unfortunately, this would still change the resolution and consequently the `uv.lock` file though. I think we may need a different concept for your use-case. What do the dependencies look like in the online case? e.g., are they on a package index? or path dependencies on a NFS?

**Comment by Ruhrpottpatriot** [comment]
> What do the dependencies look like in the online case? e.g., are they on a package index? or path dependencies on a NFS?

When working in-house they're included either from an internal package index, or git repository (in case we're still in the process of developing them). Both package index and git are internal only, this is a non negotiable requirement from our security department.

When we are in the field we use provisioned devices for that purpose and which get wiped after we're finished. At the moment we're cloning the required repositories to each device and then set `[tool.uv.sources]` to the local path of each package manually. For longer excursions we _may_ have an NFS, but usually this isn't the case. Sometimes we also have packages that come from project-partners which obviously are not giving us access to their internal network and usually give us code as a folder on a USB-Stick.

**Comment by zanieb** [comment]
So for the Git dependencies, I presume the local path is just a checkout of the repository. For the index dependencies, do you similarly use a local clone of the source tree? Or do you put the sdist / wheel in a local directory?

**Comment by Ruhrpottpatriot** [comment]
Usually it's a clone of the source tree for every (internal) dependency we have. I know that if we had wheels/sdists we could use a flat package index and iirc we did this in the past. However, we found that sometimes we have to extend/change packages that we internally pull from the index. That's very rare, but it happens occasionally. In-house this is no problem as we can change the packages and push a new version quickly, but since we don't have any access to the internal network (not even through a VPN) this isn't possible, so we usually clone the source for every internal package.

**Comment by zanieb** [comment]
Yeah I was hoping we could just add a "mirror" concept for indexes which would let you fetch the artifact from a local directory instead of the remote index, but if you're changing to source trees that's not going to work.

**Comment by ccurreri** [comment]
Also have this use case. We develop lots of loosely coupled packages that get developed simultaneously. This results in local git copies being installed in editable mode on our dev machines, but at potentially different system paths. Outside active local development the "correct" source of the packages is a local package index.

This leads to an awkward situation where I personally want to source the file from my disk but most other cases should source it from the package index. This would be neatly solved if I could specify sources without having to check that in.

In other words, I don't want to check in a pyproject.toml with local paths that only work on my system. I'd much prefer our pyproject.toml reference the index (since that's what ci/cd will build from) and then override it on my system.

Or more concretely:

I'm the principal developer of pkgA and it uses pkgB and pkgC. I also maintain pkgB. Coworker (or build system etc) only works on pkgA. I want to locally source pkgB since I need to make frequent changes to it, but coworker doesn't and just sources it from the package index.

**Comment by oxysoft** [comment]
Still hitting this in 2026 in a way that causes real project breakage, not just config confusion.

Concrete workflow:

- shared project depends on another package
- some developers need that dependency from a local checkout on disk
- shared `pyproject.toml` must remain portable for CI / fresh clones / external maintainers

The natural solution is an untracked project-local `uv.toml` with `sources`, but `uv` still rejects that. The result is that teams are pushed toward either:

- committing machine-specific paths in `pyproject.toml`, or
- using repeated CLI-only workarounds like `uv run --with-editable 'pkg @ file:///...' ...`

We hit both sides of this:

```text
error: Distribution not found at: file:///home/nuck/insiqui/imgui_vibekit
```

and then:

```text
error: Failed to parse: `uv.toml`. The `sources` field is not allowed in a `uv.toml` file. `sources` is only applicable in the context of a project, and should be placed in a `pyproject.toml` file instead.
```

Cross-linking the two related open workflow issues because they describe the same missing capability from different angles:

- #12013
- #11632

**Comment by ahormazabal-strider** [comment]
Yes we have a similar issue ourselves. We keep an internal pypi server, and for performance reasons we need to specify sources so some libraries go straight to the internal pipy. So when i want to develop with a local editable copy of one of the libraries i'm forced to change my pyproject.toml and be careful not to commit it.
